bcc/module.go

```diff
@@ -274,6 +288,46 @@ func (bpf *Module) AttachTracepoint(name string, fd int) error {
 	return nil
 }
 
+// AttachPerfEvent attaches a perf event fd to a function
+// Argument 'evType' is a member of 'perf_type_id' enum in the kernel
+// header 'include/uapi/linux/perf_event.h'. Argument 'evConfig'
+// is one of PERF_COUNT_* constants in the same file.
+func (bpf *Module) AttachPerfEvent(evType, evConfig int, samplePeriod int, sampleFreq int, pid, cpu, groupFd, fd int) error {
+	key := fmt.Sprintf("%d:%d", evType, evConfig)
+	if _, ok := bpf.perfEvents[key]; ok {
+		return nil
+	}
+
+	res := make([]int, runtime.NumCPU())
+
+	if cpu > 0 {
+		r, err := C.bpf_attach_perf_event(C.int(fd), C.uint(evType), C.uint(evConfig), C.ulong(samplePeriod), C.ulong(sampleFreq), C.int(pid), C.int(cpu), C.int(groupFd))
+		if r < 0 {
+			return fmt.Errorf("failed to attach BPF perf event: %v", err)
+		}
+
+		res[cpu] = int(r)
+	} else {
+		cpus, err := cpuonline.Get()
+		if err != nil {
+			return fmt.Errorf("failed to determine online cpus: %v", err)
+		}
+
+		for _, i := range cpus {
+			r, err := C.bpf_attach_perf_event(C.int(fd), C.uint(evType), C.uint(evConfig), C.ulong(samplePeriod), C.ulong(sampleFreq), C.int(pid), C.int(i), C.int(groupFd))
+			if r < 0 {
+				return fmt.Errorf("failed to attach BPF perf event: %v", err)
+			}
+
+			res[i] = int(r)
```

Comment:
In a scenario where CPUs `[0, 2, 3]` are online, this leads to a `panic: runtime error: index out of range` as r[0] - r[2] are allocated. So I think this needs to be changed to `res = append(res, int(r))`.

Reply:
Nice catch, updated.